util: remove Boost posix_time usage from GetTime*

fanquake opened the pull request:
I have a followup that should remove the last of our `boost:posix_time` usage in `ParseISO8601DateTime`, but that will likely need more cross-platform testing/discussion, so have just split them up as this change is straight forward.

theuni committed:
time: add runtime sanity check

std::chrono::system_clock.time_since_epoch and time_t(0) are not guaranteed
to use the Unix epoch timestamp, but in practice they almost certainly will.
Any differing behavior will be assumed to be an error, unless certain
platforms prove to consistently deviate, at which point we'll cope with it
by adding offsets.

Do a quick runtime check to verify that
time_t(0) == std::chrono::system_clock's epoch time == unix epoch.

Co-authored-by: Anthony Towns <aj@erisian.com.au>

Commit:
util: Use std::chrono for time getters

MarcoFalke commented:
Duplicate of #18282?

fanquake commented:
> Duplicate of #18282?

Do you want to reopen that and include the epoch assertion/sanity check? Otherwise I can update this PR.

MarcoFalke commented:
My plan was to wait till c++20, because it means less code to write.

fanquake commented:
> My plan was to wait till c++20, because it means less code to write.

I'm not sure that waiting the many years until the project will actually be able to use C++20, while preventing the removal of Boost, and use of std library features is worth not having to write a few additional lines of code. Especially since the assumption we have to make here is just a Unix time epoch. If it was something more controversial, or involved having to write hundreds of lines of additional code, which would incur a high maintenance overhead, I might agree, but that shouldn't be the case here.

MarcoFalke commented:
Obviously no objection to adding the sanity check now and removing it when we switch to c++20. I just meant to say that I am not going to write (rebase) it.

laanwj commented:
Concept ACK

Let's go with a sanity check. I know I disagreed on this in https://github.com/bitcoin/bitcoin/pull/9566/files#r96350569 but as we age we get tired of the same old things and just want to move forward.

practicalswift commented:
With the simple runtime sanity check from #9566 we're guaranteed that we get UTC.

Concept ACK (assuming the sanity check is added of course)

ajtowns commented:
Concept ACK; agree with adding the sanity check and not waiting for C++20.

fanquake commented:
I've changed this to pull in @theuni's sanity check (with some adjustments by @ajtowns , and some review comments from #9566 addressed). I've also dropped my commit in favour of @MarcoFalke  change from #18282.

practicalswift commented:
Tested ACK 9266f7497f256d780178829e0f3a29ddaeb794ba

Startup fails as expected if the sanity doesn't pass:

```
$ src/bitcoind
Error: Clock epoch mismatch. Aborting.
Error: Initialization sanity check failed. Bitcoin Core is shutting down.
```

After this PR is merged the only remaining user of `boost::posix_time` is `ParseISO8601DateTime`.

@elichai wrote a Boost-free replacement for `ParseISO8601DateTime` (then called `DecodeDumpTime`) two years ago in #17245. That replacement could be re-submitted as a new PR to allow for permanent removal of the `boost/date_time/posix_time/posix_time.hpp` dependency 🎉

laanwj commented:
Code review ACK 9266f7497f256d780178829e0f3a29ddaeb794ba

hebasto commented:
@fanquake 
> I have a followup that should remove the last of our `boost:posix_time` usage in `ParseISO8601DateTime`...

> I'm not sure that waiting the many years until the project will actually be able to use C++20...

As `std::chrono::parse` is available now, do you mind submitting your follow up? If not, I can do it :)
